test/support/warning_filter.rb

```diff
@@ -0,0 +1,53 @@
+require 'pathname'
+require 'set'
+
+module RegexpParserTest
+  # Stream filter for warnings
+  class WarningFilter
+    PATTERN = /\A(?:.+):(?:\d+): warning: (?:.+)\n\z/
+
+    # Use warning filter during block execution
+    #
+    # @return [Array<String>]
+    def self.use
+      original = $stderr
+      $stderr = filter = new(original)
+      yield
+      filter.warnings
+    ensure
+      $stderr = original
+    end
+
+    # Initialize object
+    #
+    # @param [#write] target
+    #
+    # @return [undefined]
+    def initialize(target)
+      @target   = target
+      @warnings = Set.new
+    end
+
+    # Write message to target filtering warnings
+    #
+    # @param [String] message
+    #
+    # @return [self]
+    def write(message)
+      return target.write(message) if PATTERN !~ message
+
+      @warnings = warnings + Set.new([message.chomp])
```

Comment:
@dkubb I usually try to prefer this sort of mutable update since it at least makes it harder for outside mutation to infect object state:

```ruby
require 'set'

class Thing1
  def initialize
    @items = Set.new
  end

  attr_reader :items

  def add(item)
    @items << item
  end
end

class Thing2
  def initialize
    @items = Set.new
  end

  attr_reader :items

  def add(item)
    @items = @items + [item]
  end
end

thing1 = Thing1.new

thing1.add(1)
set1 = thing1.items
thing1.add(2)
set1 << 'external mutation'

thing1.items # => #<Set: {1, 2, "external mutation"}>
set1         # => #<Set: {1, 2, "external mutation"}>

thing2 = Thing2.new
thing2.add(1)
set2 = thing2.items
thing2.add(2)
set2 << 'external mutation'

thing2.items # => #<Set: {1, 2}>
set2         # => #<Set: {1, "external mutation"}>
```